File: leet/P95.py
#!/usr/bin/env python

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):

    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def node(val, left, right):
            node = TreeNode(node)
            node.left = left
            node.right = right
            return node

        def trees(start, end):
            return [node(x, left, right)
                    for x in range(start, end+1)
                    for left in trees(start, x-1)
                    for right in trees(x+1, end)]
    return trees(1, n)

        
if __name__ == "__main__":

File: leet/P95.cpp
#include <cstdio>  
#include <cmath>
#include <climits>
#include <cstdlib>
#include <algorithm>
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>   // std:string std:stoi

using namespace std;

void printVector(vector<int>& v) {
    cout << "[ ";
    for (auto x:v)
        cout << x << " ";
    cout <<"] " << endl;
}

vector<TreeNode*> create(int start, int end) {
    vector<TreeNode*> trees;
    if (start > end) {
        trees.push_back(NULL);
        return trees;
    }
    for (int x=start; x<=end; ++x) 
        for (auto left: create(start, x-1))
            for (auto right: create(x+1, end)) {
                TreeNode* root = new TreeNode(x);
                root->left = left;
                root->right = right;
                trees.push_back(root);
            }
    return trees; 
}

vector<TreeNode*> generateTrees(int n) {
    vector<TreeNode*> res;
    if (n != 0)
        res = create(1, n);
    return res;
}

int main(int argc, char** argv) {
    return 0;
}
